File: service/src/main/java/no/nav/melosys/service/dokument/brev/mapper/AvslagYrkesaktivMapper.java
package no.nav.melosys.service.dokument.brev.mapper;

import java.util.Set;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.datatype.DatatypeConfigurationException;

import no.nav.dok.melosysbrev._000081.BrevdataType;
import no.nav.dok.melosysbrev._000081.Fag;
import no.nav.dok.melosysbrev._000081.LovvalgsperiodeType;
import no.nav.dok.melosysbrev._000081.ObjectFactory;
import no.nav.dok.melosysbrev.felles.melosys_felles.*;
import no.nav.melosys.domain.*;
import no.nav.melosys.domain.behandling.Behandling;
import no.nav.melosys.domain.behandling.Behandlingsresultat;
import no.nav.melosys.domain.kodeverk.Anmodningsperiodesvartyper;
import no.nav.melosys.domain.kodeverk.Sakstyper;
import no.nav.melosys.domain.kodeverk.begrunnelser.Art16_1_avslag;
import no.nav.melosys.exception.TekniskException;
import no.nav.melosys.service.dokument.brev.BrevData;
import no.nav.melosys.service.dokument.brev.BrevDataAvslagYrkesaktiv;
import org.apache.commons.lang.StringUtils;
import org.xml.sax.SAXException;

import static no.nav.melosys.domain.kodeverk.Vilkaar.*;
import static no.nav.melosys.service.dokument.brev.mapper.felles.BrevMapperUtils.convertToXMLGregorianCalendarRemoveTimezone;
import static no.nav.melosys.service.dokument.brev.mapper.felles.VilkaarbegrunnelseFactory.*;

public class AvslagYrkesaktivMapper implements BrevDataMapper {

    private static final String XSD_LOCATION = "melosysbrev/melosys_000081.xsd";
    private static final String JA = "true";

    @Override
    public String mapTilBrevXML(FellesType fellesType, MelosysNAVFelles navFelles, Behandling behandling, Behandlingsresultat resultat,
                                BrevData brevDataFelles) throws JAXBException, SAXException, TekniskException {
        BrevDataAvslagYrkesaktiv brevdata = (BrevDataAvslagYrkesaktiv) brevDataFelles;
        Fag fag = mapFag(behandling, resultat, brevdata);

        if (brevdata.anmodningsperiodeSvar.isPresent()) {
            brevdata.anmodningsperiodeSvar.ifPresent(aps ->
                mapArt161AvslagFraAnmodningsperiode(fag, aps));
        } else {
            mapArt161Avslag(fag, brevdata);
        }

        JAXBElement<BrevdataType> brevdataTypeJAXBElement = mapintoBrevdataType(fellesType, navFelles, fag);
        return JaxbHelper.marshalAndValidate(brevdataTypeJAXBElement, XSD_LOCATION);
    }

    private void mapArt161AvslagFraAnmodningsperiode(Fag fag, AnmodningsperiodeSvar svar) {
        fag.setBegrunnelseFritekst(svar.getBegrunnelseFritekst());
        Art161AvslagBegrunnelse avslagBegrunnelse = lagTomArt161AvslagBegrunnelse();
        fag.setArt161AvslagBegrunnelse(avslagBegrunnelse);
    }

    private Fag mapFag(Behandling behandling, Behandlingsresultat resultat, BrevDataAvslagYrkesaktiv brevData) throws TekniskException {
        Fag fag = new Fag();

        if (behandling.getFagsak().getType() == Sakstyper.EU_EOS) {
            // Respons fra regelmodulen skiller ikke mellom begrunnelser for 883/2004 (MELOSYS-1863)
            fag.setInngangsvilkårBegrunnelse(InngangsvilkaarBegrunnelseKode.EOS_BORGER);
        } else {
            throw new TekniskException("Forholdet er ikke dekket av inngangsvilkårene for 883/2004");
        }

        fag.setForetakNavn(brevData.hovedvirksomhet.navn);
        // Frilansaktivitet håndteres ikke i Lev 1
        fag.setYrkesaktivitet(YrkesaktivitetsKode.fromValue(brevData.yrkesaktivitet.getKode()));

        fag.setArbeidsland(brevData.arbeidsland);
        fag.setLovvalgsperiode(lagLovvalgsperiodeType(resultat));

        Set<VilkaarBegrunnelse> art121Begrunnelser = resultat.hentVilkaarbegrunnelser(FO_883_2004_ART12_1);
        fag.setArt121Begrunnelse(mapArt121BegrunnelseType(art121Begrunnelser));

        Set<VilkaarBegrunnelse> art121ForutgåendeBegrunnelser = resultat.hentVilkaarbegrunnelser(ART12_1_FORUTGAAENDE_MEDLEMSKAP);
        fag.setArt121ForutgåendeBegrunnelse(mapArt121ForutgaaendeBegrunnelseType(art121ForutgåendeBegrunnelser));

        Set<VilkaarBegrunnelse> art122Begrunnelser = resultat.hentVilkaarbegrunnelser(FO_883_2004_ART12_2);
        fag.setArt122Begrunnelse(mapArt122BegrunnelseType(art122Begrunnelser));

        Set<VilkaarBegrunnelse> art122NormalVirksomhetBegrunnelse = resultat.hentVilkaarbegrunnelser(ART12_2_NORMALT_DRIVER_VIRKSOMHET);
        fag.setArt122NormalVirksomhetBegrunnelse(mapArt122NormalVirksomhetBegrunnelseType(art122NormalVirksomhetBegrunnelse));

        fag.setFritekst(brevData.fritekst);

        brevData.anmodningsperiodeSvar
            .map(AnmodningsperiodeSvar::getAnmodningsperiodeSvarType)
            .map(Anmodningsperiodesvartyper::getKode)
            .map(AnmodningsPeriodeSvarTypeKode::valueOf)
            .ifPresent(fag::setAnmodningsPeriodeSvarType);

        if (brevData.erArt16UtenArt12) {
            fag.setArt16UtenArt12(JA);
        }

        return fag;
    }

    void mapArt161Avslag(Fag fag, BrevDataAvslagYrkesaktiv brevdata) throws TekniskException {
        Vilkaarsresultat vilkaarsresultat = brevdata.art16Vilkaar;
        Set<VilkaarBegrunnelse> art161Begrunnelser = vilkaarsresultat.getBegrunnelser();
        Art161AvslagBegrunnelse art161AvslagBegrunnelser = lagTomArt161AvslagBegrunnelse();
        for (VilkaarBegrunnelse vilkaarBegrunnelse : art161Begrunnelser) {
            Art16_1_avslag artikkel161AvslagKode = Art16_1_avslag.valueOf(vilkaarBegrunnelse.getKode());
            switch (artikkel161AvslagKode) {
                case OVER_5_AAR:
                    art161AvslagBegrunnelser.setOver5Aar(JA);
                    break;
                case INGEN_SPESIELLE_FORHOLD:
                    art161AvslagBegrunnelser.setIngenSpesielleForhold(JA);
                    break;
                case SAERLIG_AVSLAGSGRUNN:
                    art161AvslagBegrunnelser.setSaerligAvslagsgrunn(JA);
                    fag.setBegrunnelseFritekst(validerFritekstbegrunnelse(vilkaarsresultat.getBegrunnelseFritekst()));
                    break;
                case SOEKT_FOR_SENT:
                    art161AvslagBegrunnelser.setSoektForSent(JA);
                    break;
                default:
                    throw new TekniskException(artikkel161AvslagKode + " støttes ikke.");
            }
        }

        fag.setArt161AvslagBegrunnelse(art161AvslagBegrunnelser);
    }

    private static Art161AvslagBegrunnelse lagTomArt161AvslagBegrunnelse() {
        return Art161AvslagBegrunnelse.builder().withIngenSpesielleForhold("")
            .withOver5Aar("")
            .withSaerligAvslagsgrunn("")
            .withSoektForSent("").build();
    }

    static String validerFritekstbegrunnelse(String begrunnelse) throws TekniskException {
        if (!StringUtils.isEmpty(begrunnelse)) {
            return begrunnelse;
        } else {
            throw new TekniskException("Ingen fritekstbegrunnelse satt for Artikkel 16.1");
        }
    }

    private static LovvalgsperiodeType lagLovvalgsperiodeType(Behandlingsresultat resultat) throws TekniskException {
        Lovvalgsperiode lovvalgsperiode = resultat.hentValidertLovvalgsperiode();
        LovvalgsperiodeType lovvalgsperiodeType = new LovvalgsperiodeType();

        try {
            lovvalgsperiodeType.setFomDato(convertToXMLGregorianCalendarRemoveTimezone(lovvalgsperiode.getFom()));
            lovvalgsperiodeType.setTomDato(convertToXMLGregorianCalendarRemoveTimezone(lovvalgsperiode.getTom()));
        } catch (DatatypeConfigurationException e) {
            throw new TekniskException(e);
        }
        return lovvalgsperiodeType;
    }

    private static JAXBElement<BrevdataType> mapintoBrevdataType(FellesType fellesType, MelosysNAVFelles navFelles, Fag fag) {
        ObjectFactory factory = new ObjectFactory();
        BrevdataType brevdataType = factory.createBrevdataType();
        brevdataType.setFelles(fellesType);
        brevdataType.setNAVFelles(navFelles);
        brevdataType.setFag(fag);
        return factory.createBrevdata(brevdataType);
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/dokument/brev/mapper/AvslagYrkesaktivMapper.kt
package no.nav.melosys.service.dokument.brev.mapper


import jakarta.xml.bind.JAXBElement
import no.nav.dok.melosysbrev._000081.BrevdataType
import no.nav.dok.melosysbrev._000081.Fag
import no.nav.dok.melosysbrev._000081.LovvalgsperiodeType
import no.nav.dok.melosysbrev._000081.ObjectFactory
import no.nav.dok.melosysbrev.felles.melosys_felles.*
import no.nav.melosys.domain.AnmodningsperiodeSvar
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Behandlingsresultat
import no.nav.melosys.domain.kodeverk.Sakstyper
import no.nav.melosys.domain.kodeverk.Vilkaar.*
import no.nav.melosys.domain.kodeverk.begrunnelser.Avslag_anmodning_begrunnelser
import no.nav.melosys.exception.TekniskException
import no.nav.melosys.service.dokument.brev.BrevData
import no.nav.melosys.service.dokument.brev.BrevDataAvslagYrkesaktiv
import no.nav.melosys.service.dokument.brev.mapper.felles.BrevMapperUtils.convertToXMLGregorianCalendarRemoveTimezone
import no.nav.melosys.service.dokument.brev.mapper.felles.VilkaarbegrunnelseFactory.*


open class AvslagYrkesaktivMapper : BrevDataMapper {

    override fun mapTilBrevXML(
        fellesType: FellesType,
        navFelles: MelosysNAVFelles,
        behandling: Behandling,
        resultat: Behandlingsresultat,
        brevDataFelles: BrevData
    ): String {
        val brevdata = brevDataFelles as BrevDataAvslagYrkesaktiv
        val fag = mapFag(behandling, resultat, brevdata)

        if (brevdata.anmodningsperiodeSvar != null) {
            mapArt161AvslagFraAnmodningsperiode(fag, brevdata.anmodningsperiodeSvar!!)
        } else {
            mapArt161Avslag(fag, brevdata)
        }

        val brevdataTypeJAXBElement = mapintoBrevdataType(fellesType, navFelles, fag)
        return JaxbHelper.marshalAndValidate(brevdataTypeJAXBElement, XSD_LOCATION)
    }

    private fun mapArt161AvslagFraAnmodningsperiode(fag: Fag, svar: AnmodningsperiodeSvar) {
        fag.begrunnelseFritekst = svar.begrunnelseFritekst
        fag.art161AvslagBegrunnelse = lagTomArt161AvslagBegrunnelse()
    }

    private fun mapFag(
        behandling: Behandling,
        resultat: Behandlingsresultat,
        brevData: BrevDataAvslagYrkesaktiv
    ): Fag = Fag().apply {
        if (behandling.fagsak.type == Sakstyper.EU_EOS) {
            inngangsvilkårBegrunnelse = InngangsvilkaarBegrunnelseKode.EOS_BORGER
        } else {
            throw TekniskException("Forholdet er ikke dekket av inngangsvilkårene for 883/2004")
        }

        foretakNavn = brevData.hovedvirksomhet?.navn
        yrkesaktivitet = YrkesaktivitetsKode.fromValue(brevData.yrkesaktivitet?.kode)
        arbeidsland = brevData.arbeidsland
        lovvalgsperiode = lagLovvalgsperiodeType(resultat)

        art121Begrunnelse = mapArt121BegrunnelseType(
            resultat.hentVilkaarbegrunnelser(
                FO_883_2004_ART12_1,
                KONV_EFTA_STORBRITANNIA_ART14_1,
                KONV_EFTA_STORBRITANNIA_ART16_1
            )
        )
        art121ForutgåendeBegrunnelse = mapArt121ForutgaaendeBegrunnelseType(resultat.hentVilkaarbegrunnelser(FORUTGAAENDE_MEDLEMSKAP))
        art122Begrunnelse = mapArt122BegrunnelseType(
            resultat.hentVilkaarbegrunnelser(
                FO_883_2004_ART12_2,
                KONV_EFTA_STORBRITANNIA_ART14_2,
                KONV_EFTA_STORBRITANNIA_ART16_3
            )
        )
        art122NormalVirksomhetBegrunnelse = mapArt122NormalVirksomhetBegrunnelseType(resultat.hentVilkaarbegrunnelser(NORMALT_DRIVER_VIRKSOMHET))

        fritekst = brevData.fritekst

        anmodningsPeriodeSvarType = brevData.anmodningsperiodeSvar?.anmodningsperiodeSvarType?.let {
            AnmodningsPeriodeSvarTypeKode.valueOf(it.kode)
        }

        if (brevData.art16UtenArt12) {
            art16UtenArt12 = JA
        }
    }

    fun mapArt161Avslag(fag: Fag, brevdata: BrevDataAvslagYrkesaktiv) {
        val vilkaarsresultat = brevdata.art16Vilkaar
        val art161Begrunnelser = vilkaarsresultat?.begrunnelser
        val art161AvslagBegrunnelser = lagTomArt161AvslagBegrunnelse()
        art161Begrunnelser?.forEach { vilkaarBegrunnelse ->
            val begrunnelseKode = Avslag_anmodning_begrunnelser.valueOf(vilkaarBegrunnelse.kode)
            when (begrunnelseKode) {
                Avslag_anmodning_begrunnelser.OVER_5_AAR -> art161AvslagBegrunnelser.over5Aar = JA
                Avslag_anmodning_begrunnelser.INGEN_SPESIELLE_FORHOLD -> art161AvslagBegrunnelser.ingenSpesielleForhold = JA
                Avslag_anmodning_begrunnelser.SAERLIG_AVSLAGSGRUNN -> {
                    art161AvslagBegrunnelser.saerligAvslagsgrunn = JA
                    fag.begrunnelseFritekst = validerFritekstbegrunnelse(vilkaarsresultat.begrunnelseFritekst)
                }

                Avslag_anmodning_begrunnelser.SOEKT_FOR_SENT -> art161AvslagBegrunnelser.soektForSent = JA
                else -> throw TekniskException("$begrunnelseKode støttes ikke.")
            }
        }

        fag.art161AvslagBegrunnelse = art161AvslagBegrunnelser
    }

    private fun lagTomArt161AvslagBegrunnelse(): Art161AvslagBegrunnelse =
        Art161AvslagBegrunnelse().withIngenSpesielleForhold("")
            .withOver5Aar("")
            .withSaerligAvslagsgrunn("")
            .withSoektForSent("")

    private fun validerFritekstbegrunnelse(begrunnelse: String?): String {
        if (!begrunnelse.isNullOrEmpty()) {
            return begrunnelse
        } else {
            throw TekniskException("Ingen fritekstbegrunnelse satt for Artikkel 16.1")
        }
    }

    private fun lagLovvalgsperiodeType(resultat: Behandlingsresultat): LovvalgsperiodeType {
        val lovvalgsperiode = resultat.hentLovvalgsperiode()
        val lovvalgsperiodeType = LovvalgsperiodeType()

        lovvalgsperiodeType.fomDato = convertToXMLGregorianCalendarRemoveTimezone(lovvalgsperiode.hentFom())
        lovvalgsperiodeType.tomDato = convertToXMLGregorianCalendarRemoveTimezone(lovvalgsperiode.tom)

        return lovvalgsperiodeType
    }

    private fun mapintoBrevdataType(
        fellesType: FellesType,
        navFelles: MelosysNAVFelles,
        fag: Fag
    ): JAXBElement<BrevdataType> {
        val factory = ObjectFactory()
        val brevdataType = factory.createBrevdataType().apply {
            this.felles = fellesType
            this.navFelles = navFelles
            this.fag = fag
        }
        return factory.createBrevdata(brevdataType)
    }

    companion object {
        private const val XSD_LOCATION = "melosysbrev/melosys_000081.xsd"
        private const val JA = "true"
    }
}
